File: backend/src/main/java/com/backend/global/jwt/JwtTokenProvider.java
/*
package com.backend.global.jwt;

import com.backend.domain.member.entity.Role;
import com.backend.global.security.CustomUserDetails;
import com.backend.global.security.CustomUserDetailsService;
import io.jsonwebtoken.*;
import io.jsonwebtoken.security.Keys;
import jakarta.annotation.PostConstruct;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.util.Date;
import java.util.List;

@Getter
@Component
@Slf4j
public class JwtTokenProvider {

    @Autowired
    private CustomUserDetailsService customUserDetailsService;

    @Value("${custom.jwt.secret-key}")
    private String secretKey;

    @Value("${custom.jwt.access-token.expire-time}")
    private long ACCESS_TOKEN_EXPIRE_TIME;

    @Value("${custom.jwt.refresh-token.expire-time}")
    private long REFRESH_TOKEN_EXPIRE_TIME;

    private SecretKey key;

    @PostConstruct
    public void init() {
        this.key = Keys.hmacShaKeyFor(secretKey.getBytes(StandardCharsets.UTF_8));
    }

    // Access Token 생성 - 로그인 후 인증
    public String generateAccessToken(Long memberId, Role role) {
        return generateToken(memberId, role, ACCESS_TOKEN_EXPIRE_TIME, "access");
    }

    // Refresh Token 생성 - Access 만료 시 재발급
    public String generateRefreshToken(Long memberId, Role role) {
        return generateToken(memberId, role, REFRESH_TOKEN_EXPIRE_TIME, "refresh");
    }

    // 공통 토큰 생성 로직
    private String generateToken(Long memberId, Role role, long expireTime, String type) {
        Date now = new Date();
        Date expiry = new Date(now.getTime() + expireTime);

        return Jwts.builder()
                .subject(String.valueOf(memberId))
                .issuedAt(now)
                .expiration(expiry)
                .claim("role", role.name())
                .claim("type", type)
                .signWith(key)
                .compact();
    }

    // 토큰 유효성 검증 (서명 검증 + 만료 체크)
    public TokenStatus validateTokenStatus(String token) {

        try {
            Jwts.parser().verifyWith(key).build().parse(token);
            return TokenStatus.VALID;
        } catch (ExpiredJwtException e) {
            log.info("== 토큰이 만료되었습니다 ==");
            return TokenStatus.EXPIRED;
        } catch (JwtException | IllegalArgumentException e) {
            log.info("== 유효하지 않은 토큰입니다 ==");
            return TokenStatus.INVALID;
        }
    }

    */
/** 토큰 파싱 관련 (Claims 읽기) *//*


    // Claims 파싱 (JWT → 내부 데이터 복호화 + 서명 검증)
    private Claims parseClaims(String token) {
        // parseSignedClaims(): JWT를 서명 검증 후 Payload(Claims) 반환
        return Jwts.parser()
                .verifyWith(key)
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }

    // 토큰에서 memberId(subject) 추출
    public Long getMemberIdFromToken(String token) {
        // subject는 String으로 저장되어 있으므로 Long으로 변환
        return Long.valueOf(parseClaims(token).getSubject());
    }

    // 토큰에서 role 추출
    public Role getRoleFromToken(String token) {
        Object role = parseClaims(token).get("role");
        return Role.valueOf(role.toString());
    }

    // JWT → Authentication으로 변환 (스프링 시큐리티는 “Authentication 객체”를 기준으로 사용자 인증 여부를 판단)
    public Authentication getAuthentication(String token) {
        Long memberPk = getMemberIdFromToken(token);
        Role role = getRoleFromToken(token);

        // DB에서 Member 엔티티 조회 후 CustomUserDetails 생성
        CustomUserDetails userDetails =
                (CustomUserDetails) customUserDetailsService.loadUserById(memberPk);

        return new UsernamePasswordAuthenticationToken(
                userDetails,  // principal을 CustomUserDetails로
                null,
                userDetails.getAuthorities()
        );
    }

    // type 조회용 메서드 추가
    public String getTokenType(String token) {
        Object type = parseClaims(token).get("type");
        return type != null ? type.toString() : "unknown";
    }

    */
/** 토큰 만료시간 조회 (쿠키 설정 등에서 사용) *//*


    public long getAccessTokenExpireTime() {
        return ACCESS_TOKEN_EXPIRE_TIME / 1000; // 초 단위로 변환 (쿠키에서 사용하기 좋음)
    }

    public long getRefreshTokenExpireTime() {
        return REFRESH_TOKEN_EXPIRE_TIME / 1000;
    }
}
*/

File: backend/src/main/java/com/backend/global/security/jwt/JwtTokenProvider.kt
package com.backend.global.security.jwt

import com.backend.domain.member.entity.Provider
import com.backend.domain.member.entity.Role
import com.backend.global.security.oauth.dto.OAuth2TempClaims
import com.backend.global.security.user.CustomUserDetails
import com.backend.global.security.user.CustomUserDetailsService
import io.jsonwebtoken.Claims
import io.jsonwebtoken.ExpiredJwtException
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.Authentication
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.util.*
import javax.crypto.SecretKey
import jakarta.annotation.PostConstruct
import java.lang.IllegalArgumentException

@Component
class JwtTokenProvider(
    private val customUserDetailsService: CustomUserDetailsService,

    @Value("\${custom.jwt.secret-key}")
    private val secretKey: String,

    @Value("\${custom.jwt.access-token.expire-time}")
    private val accessTokenExpireTimeMs: Long,

    @Value("\${custom.jwt.refresh-token.expire-time}")
    private val refreshTokenExpireTimeMs: Long
) {

    private lateinit var key: SecretKey

    @PostConstruct
    fun init() {
        key = Keys.hmacShaKeyFor(secretKey.toByteArray(StandardCharsets.UTF_8))
    }

    /** Access Token 생성 */
    fun generateAccessToken(memberId: Long, role: Role): String {
        return generateToken(
            subject = memberId.toString(),
            claims = mapOf(
                "role" to role.name,
                "type" to TokenType.ACCESS.name
            ),
            expireTime = accessTokenExpireTimeMs
        )
    }

    /** Refresh Token 생성 */
    fun generateRefreshToken(memberId: Long, role: Role): String {
        return generateToken(
            subject = memberId.toString(),
            claims = mapOf(
                "role" to role.name,
                "type" to TokenType.REFRESH.name
            ),
            expireTime = refreshTokenExpireTimeMs
        )
    }

    /** ⭐ 소셜 회원가입 임시 토큰 (provider / providerId / email 기반) */
    fun generateOAuth2TempToken(
        provider: Provider,
        providerId: String,
        email: String
    ): String {
        return generateToken(
            subject = providerId,
            claims = mapOf(
                "provider" to provider.name,
                "providerId" to providerId,
                "email" to email,
                "type" to TokenType.OAUTH2_TEMP.name
            ),
            expireTime = 1000L * 60 * 10      // 10 minutes
        )
    }

    /** 공통 토큰 생성 로직 */
    private fun generateToken(
        subject: String,
        claims: Map<String, Any>,
        expireTime: Long
    ): String {
        val now = Date()
        val expiry = Date(now.time + expireTime)

        return Jwts.builder()
            .subject(subject)
            .issuedAt(now)
            .expiration(expiry)
            .claims(claims)
            .signWith(key)
            .compact()
    }

    /** 토큰 유효성 검증 (서명 및 만료 확인) */
    fun validateTokenStatus(token: String): TokenStatus =
        try {
            Jwts.parser()
                .verifyWith(key)
                .build()
                .parse(token)
            TokenStatus.VALID
        } catch (e: ExpiredJwtException) {
            log.info("== 토큰이 만료되었습니다 ==")
            TokenStatus.EXPIRED
        } catch (e: JwtException) {
            log.info("== 유효하지 않은 토큰입니다 ==")
            TokenStatus.INVALID
        } catch (e: IllegalArgumentException) {
            log.info("== 유효하지 않은 토큰입니다 ==")
            TokenStatus.INVALID
        }

    /** Claims 파싱 (JWT Payload 디코드) */
    private fun parseClaims(token: String): Claims =
        Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(token)
            .payload

    /** 토큰에서 memberId(subject) 추출 */
    fun getMemberIdFromToken(token: String): Long =
        parseClaims(token).subject.toLong()

    /** 토큰에서 Role 추출 */
    fun getRoleFromToken(token: String): Role =
        Role.valueOf(parseClaims(token)["role"].toString())

    /** 토큰 타입 조회 (access / refresh) */
    fun getTokenType(token: String): TokenType =
        TokenType.valueOf(parseClaims(token)["type"].toString())

    /** JWT → Authentication 변환 */
    fun getAuthentication(token: String): Authentication {
        val memberPk = getMemberIdFromToken(token)
        val userDetails = customUserDetailsService.loadUserById(memberPk) as CustomUserDetails

        return UsernamePasswordAuthenticationToken(
            userDetails,
            null,
            userDetails.authorities
        )
    }

    /** 만료 시간 (초 단위 반환) */
    val accessTokenExpireTime: Long get() = accessTokenExpireTimeMs / 1000
    val refreshTokenExpireTime: Long get() = refreshTokenExpireTimeMs / 1000

    companion object {
        private val log = LoggerFactory.getLogger(JwtTokenProvider::class.java)
    }

    fun parseOAuth2TempToken(token: String): OAuth2TempClaims {
        val claims = parseClaims(token)

        return OAuth2TempClaims(
            provider = Provider.valueOf(claims["provider"].toString()),
            providerId = claims["providerId"].toString(),
            email = claims["email"].toString()
        )
    }
}